webrender/src/device.rs

```diff
@@ -1075,26 +1099,69 @@ impl Device {
         }
     }
 
-    pub fn create_fbo_for_texture_if_necessary(&mut self, texture_id: TextureId) {
-        if !self.textures.get(&texture_id).unwrap().fbo_ids.is_empty() {
-            return
-        }
+    pub fn create_fbo_for_texture_if_necessary(&mut self,
+                                               texture_id: TextureId,
+                                               layer_count: Option<i32>) {
+        let texture = self.textures.get_mut(&texture_id).unwrap();
+
+        match layer_count {
+            Some(layer_count) => {
+                debug_assert!(layer_count > 0);
 
-        let fbo_ids: Vec<_> =
-            gl::gen_framebuffers(1).into_iter().map(|fbo_id| FBOId(fbo_id)).collect();
-        for fbo_id in &fbo_ids[..] {
-            gl::bind_framebuffer(gl::FRAMEBUFFER, fbo_id.0);
+                // If we have enough layers allocated already, just use them.
+                // TODO(gw): Probably worth removing some after a while if
+                //           there is a surplus?
+                let current_layer_count = texture.fbo_ids.len() as i32;
+                if current_layer_count >= layer_count {
+                    return;
+                }
 
-            gl::framebuffer_texture_2d(gl::FRAMEBUFFER,
-                                       gl::COLOR_ATTACHMENT0,
-                                       gl::TEXTURE_2D,
-                                       texture_id.0,
-                                       0);
+                let (internal_format, gl_format) = gl_texture_formats_for_image_format(texture.format);
+                let type_ = gl_type_for_texture_format(texture.format);
+
+                gl::tex_image_3d(texture_id.target,
```

Comment:
This is concerning. I don't see the previous implementation actually reallocating the texture, like this one. Neither does the method name imply that the texture is going to be changed (`create_fbo_for_texture_if_necessary`).